executor/analyze.go

```diff
@@ -381,3 +398,425 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
+			e.scanTasks = append(e.scanTasks, loc)
+		} else {
+			e.sampLocs <- loc
+		}
+	}
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	// Do get regions row count.
+	close(e.sampLocs)
+	atomic.StoreUint64(&e.rowCount, 0)
+	needRebuildForRoutine := make([]bool, e.concurrency)
+	errs := make([]error, e.concurrency)
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		go e.getSampRegionsRowCount(bo, &needRebuildForRoutine[i], &errs[i])
+	}
+	e.wg.Wait()
+	for i := 0; i < e.concurrency; i++ {
+		if errs[i] != nil {
+			return false, errors.Trace(errs[i])
+		}
+	}
+	for i := 0; i < e.concurrency; i++ {
+		if needRebuildForRoutine[i] == true {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (e *AnalyzeFastExec) handleBatchGetResponse(resp *kvrpcpb.BatchGetResponse, collectors []*statistics.SampleCollector, cursor *int32) (err error) {
+	length := int32(len(resp.Pairs))
+	timeZone := e.ctx.GetSessionVars().Location()
+	newCursor := atomic.AddInt32(cursor, length)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	hasIdxInfo := len(e.idxsInfo)
+	for i, pair := range resp.Pairs {
+		samplePos := newCursor - length + int32(i)
+		if hasPKInfo > 0 {
+			collectors[0].Samples[samplePos].Ordinal = int(samplePos)
+			collectors[0].Samples[samplePos].Value, err = tablecodec.DecodeColumnValue(pair.Value, &e.pkInfo.FieldType, timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for j, colInfo := range e.colsInfo {
+			collectors[hasPKInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+			collectors[hasPKInfo+j].Samples[samplePos].Value, err = tablecodec.DecodeColumnValue(pair.Value, &colInfo.FieldType, timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for j, idxInfo := range e.idxsInfo {
+			collectors[len(e.colsInfo)+hasIdxInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+			idxVals := make([]types.Datum, 0, len(idxInfo.Columns))
+			for _, idxCol := range idxInfo.Columns {
+				val, err := tablecodec.DecodeColumnValue(pair.Value, &e.table.Cols()[idxCol.Offset].FieldType, timeZone)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				idxVals = append(idxVals, val)
+			}
+			var bytes []byte
+			bytes, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, bytes, idxVals...)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			collectors[len(e.colsInfo)+hasIdxInfo+j].Samples[samplePos].Value = types.NewBytesDatum(bytes)
+		}
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleScanResponse(resp *kvrpcpb.ScanResponse, collectors []*statistics.SampleCollector) (err error) {
+	timeZone := e.ctx.GetSessionVars().Location()
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	hasIdxInfo := len(e.idxsInfo)
+	for _, pair := range resp.Pairs {
+		// reservoir sampling
+		e.rowCount++
+		randNum := rand.Int63n(int64(e.rowCount))
+		if randNum > int64(maxSampleSize) {
```

Comment:
We should use the current sample size? Not max.

Reply:
Yeah, in code logic. The sample size is maxSampleSize or 0.